JS-SDK: improve types, factory functions for core contracts

Leechael opened the pull request:
- [x] Returns decoded object instead raw result for `instantiate`.
- [x] Factory functions for system contract, also typed.
- [x] Factory functions for logger contract, also typed.
- [x] Backporting the latest update from @vabaly 
- [x] Check compatibility for ink 4.0
- [ ] ~optimize for ink 4.0~

I break it into smaller tasks.

- [x] `PinkContractPromise` query
- [x] `PinkContractPromise` tx
- [x] `SystemContractPromise`
- [x] `PinkLoggerContractPromise`
- [x] `PinkCodePromise`
- [x] `PinkBlueprintPromise`

Post-design concepts tasks:
- [x] use it in e2e tests.
- [ ] ~add some test cases for `SystemContractPromise` and `PinkLoggerContractPromise`~
- [ ] ~deprecated old APIs (send warning with `console.warn`)~

For the "big breaking change" update:
- [x] Update protobuf.
- [x] Update System Contract ABI.
- [x] Use renamed substrate API.
- [x] New estimate parameters.

Leechael commented:
`PinkContractPromise` is the foundation class, so all options must be specified manually. We will soon simplify the process using inheritance.

```javascript
import { ApiPromise, wsProvider, Keyring, WsProvider } from '@polkadot/api'
import { typeDefinitions } from '@polkadot/types'
import { ready, PinkContractPromise, createPruntimeApi, types as PhalaTypes } from '@phala/sdk'

async function main() {
  await ready()

  const api = await ApiPromise.create({
    provider: new WsProvider("wss://poc5.phala.network/ws"),
    types: {
      ...PhalaTypes,
      ...typeDefinitions,
    },
    noInitWarn: true,
  })
  const keyring = new Keyring({ type: 'sr25519' });
  const pair = keyring.addFromUri("//Alice");


  //
  // grab cluster & pruntime information from chain so we don't need setup manually.
  //

  const clusters = []
  {
    const result = await api.query.phalaFatContracts.clusters.entries()
    result.forEach(([storageKey, value]) => {
      const clusterId = storageKey.toHuman()
      const clusterInfo = value.unwrap().toHuman()
      clusters[clusterId] = clusterInfo
      clusters.push([clusterId, clusterInfo])
    })
  }
  const [clusterId, clusterInfo] = clusters[0]
  const remotePubkey = clusterInfo.workers[0]
  console.log('Cluster ID:', clusterId)
  console.log('Worker:', remotePubkey)

  let pruntimeURL = ''
  {
    const result = await api.query.phalaRegistry.endpoints(clusters[clusterId].workers[0])
    pruntimeURL = R.path(['V1', '0'], result.toHuman())
  }
  console.log('PruntimeURL: ', pruntimeURL)

  //
  // Assume we interact with the fat badges contract in oracle workshop and we already known the contract ID.
  //
  // https://github.com/Phala-Network/oracle-workshop/tree/master/fat_badges
  //

  const contractId = ''

  let contractKey = ''
  {
    contractKey = (
      await api.query.phalaRegistry.contractKeys(contractId)
    ).toString();
  }

  const contract = new PinkContractPromise(api, createPruntimeApi(pruntimeURL), remotePubkey, abi, contractId,contractKey)

  // query
  const total = await contract.query.getTotalBadges(pair)
  console.log('total:', total)

  // costs estimation
  const { gasRequired, storageDeposit } = await contract.query.newBadge(pair, 'Badge01')
  console.log('gasRequired & storageDeposit: ', gasRequired, storageDeposit)

  const options = {
      gasLimit: gasRequired.refTime,
      storageDepositLimit: storageDeposit.isCharge ? storageDeposit.asCharge : null,
  }
  const extrinsicsEvents = contract.tx.newBadge(options, 'Badge01').signAndSend(pair, { nonce: -1 })
  console.log('extrinsic events: ', extrinsicsEvents)
}
```

h4x3rotab commented on frontend/packages/sdk/src/create.ts:
```diff
@@ -80,7 +81,7 @@ export interface CreateFnOptions {
   api: ApiPromise;
   sidevmQuery: SidevmQuery;
-  instantiate: SidevmQuery;
+  instantiate: SidevmQuery<ContractInstantiateResult | InkQueryError>;
```
Should we rename `SidevmQuery` to `RawQuery` and alias it as `SidevmQuery`? To me `instantiate()` is not a query to sidevm.

Leechael replied:
IMO old codes will not continue updates and leave it there for compatible reason. For TypeScript project type declaration update may break downstream's linting.

h4x3rotab commented on frontend/packages/sdk/src/create.ts:
```diff
@@ -101,34 +102,50 @@ export interface ClusterInfo {
+
+export interface InkResponse extends IMap {
+  nonce: Vec<u8>;
+  result: Result<InkQueryOk, InkQueryError>;
```
With ink! 4.0, not sure if the return result is still `Result<T, DispatchError>` or `Result<Result<T, DispatchError>, InkLangError>`

Leechael replied:
> With ink! 4.0, not sure if the return result is still `Result<T, DispatchError>` or `Result<Result<T, DispatchError>, InkLangError>`

Should be `Result<Result<T, DispatchError>, InkLangError>` but not going to handle that in legacy codes.

h4x3rotab replied:
Cool. Looking forward to retire the legacy code.

h4x3rotab commented on frontend/packages/sdk/src/contracts/PinkContract.ts:
```diff
@@ -0,0 +1,287 @@
+      : new Abi(abi, api.registry.getChainProperties());
+    this.api = api;
+    this._decorateMethod = toPromiseMethod;
+    this._isWeightV1 = !api.registry.createType<WeightV2>('Weight').proofSize;
```
We don't need to support WeightV1

h4x3rotab commented on frontend/packages/sdk/src/contracts/PinkContract.ts:
```diff
@@ -0,0 +1,287 @@
+        InkMessage: {
+          nonce: hexAddPrefix(randomHex(32)),
+          // FIXME: unexpected u8a prefix
+          message: api.createType("Vec<u8>", encParams).toHex(),
```
It's suggested to pass a hex string when constructing `Vec<u8>` fields in polkadot.js. So here I'd suggest to call `u8aToHex()` from polkadotjs utils lib.

Leechael replied:
I'm revert this change because I found the encoded messages decode failed in PRuntime.

h4x3rotab replied:
Which change? Btw, `api.createType("Vec<u8>", hexData).toHex() == hexData` is always true. It's suggested to use `u8aToHex(encParams)` here.

Leechael replied:
As I checked, `api.createType("Vec<u8>", hexData).toHex() != u8aToHex(encParams)`. I'm tried print these two into console:

```javascript
console.log('createType', api.createType("Vec<u8>", encParams).toHex())
console.log('u8aToHex', u8aToHex(encParams))
```

And got:
```
createType 0x0df6030d48426164676531363833333830393133333932
u8aToHex 0x5c0df6030d48426164676531363833333830393133333932
```

I guess for `Vec`, it will insert prefix `5c` for the encoded result.

h4x3rotab replied:
`5c` is the scale prefix, indicating the length of the array. Apparently polkadotjs has a weird API for `api.createType("Vec<u8>", data)`

- If `data` is u8a, apply the scale decode (considering the length prefix)
- If `data` is a hex string, create a `Vec<u8>` with its content
- If `data` is not a hex string, create a `Vec<u8>` with its utf8 encoding

To convert `Vec<u8>` to a hex, there are three ways:

- `data.toHex()`: return the content of the array
- `u8aToHex(data)`: same as before
- `u8aToHex(data.toU8a())`: return the scale encoded bytes array

h4x3rotab replied:
```javascript
let data = api.createType('Vec<u8>', '0xAABB');
console.log({
    toHex: data.toHex(),
    toU8a: u8aToHex(data.toU8a()),
    u8aToHex: u8aToHex(data),
});
// { toHex: '0xaabb', toU8a: '0x08aabb', u8aToHex: '0xaabb' }

const encodedHex = '0x08aabb';
data = api.createType('Vec<u8>', hexToU8a(encodedHex));
console.log({
    toHex: data.toHex(),
    toU8a: u8aToHex(data.toU8a()),
    u8aToHex: u8aToHex(data),
});
// { toHex: '0xaabb', toU8a: '0x08aabb', u8aToHex: '0xaabb' }
```

h4x3rotab commented on frontend/packages/sdk/src/contracts/PinkContract.ts:
```diff
@@ -0,0 +1,287 @@
+      const { debugMessage, gasConsumed, gasRequired, result, storageDeposit } = api.createType<ContractExecResult>(
+        "ContractExecResult",
+        (
+          api.createType("InkResponse", hexAddPrefix(data)).toJSON() as {
```
`data` is already a hex string with prefix?

h4x3rotab replied:
Ping

h4x3rotab commented on frontend/packages/sdk/src/contracts/PinkContract.ts:
```diff
@@ -0,0 +1,287 @@
+      this.abi.findMessage(messageOrId).toU8a(params)
+    ).withResultTransform((result: ISubmittableResult) =>
+      // ContractEmitted is the current generation, ContractExecution is the previous generation
+      new ContractSubmittableResult(result, applyOnEvent(result, ['ContractEmitted', 'ContractExecution'], (records: EventRecord[]) =>
```
Due to the difference between our execution model and the original ink, we will get no event. The original ink runs the tx on the blockchain. Once the tx is finished, it can filter out the events emitted by the tx call, and return to the user. However on Phala, contracts execute on offchain worker. It can still emit events, but we can only get it from pruntime.

It's good to leave a TODO and move forward for now. However, if we want to support the real event emission, we should:

1. Make sure the emitted event can be retrieved from the worker (pruntime api, or log server)
2. Instead of looking from the `records`, we should wait for the tx execution on worker, and get the event records from it

h4x3rotab replied:
CC @kvinwang

h4x3rotab replied:
https://github.com/Phala-Network/index-contract/pull/42#discussion_r1102239238

Leechael replied:
The `ContractSubmittableResult` will return the block hash, and we can find two on-chain events: `balances.Transfer` and `phalaFatContracts.Transfered`, so I don't think we will get no event here.

To clarify this behavior in the documentation, we should inform users that the events they see only indicate that their transaction has been submitted to the chain. They will still need to fetch the logs from PRuntime to catch any further events that may occur.

My question here: are we logging in PRuntime with the tx code hash?

h4x3rotab replied:
@kvinwang To answer.

kvinwang replied:
> are we logging in PRuntime with the tx code hash?

PRuntime doesn't log the events with the tx hash. The tx hash isn't visible there.
There is a `nonce` passed into pruntime, but it is just attached to the MessageOutput and still isn't attached to events in the log.

h4x3rotab reviewed (commented):
Additional Qs:

1. Do you plan to retire `create()` API in the future, given we already have `PinkContractPromise`?
2. Will you add `PinkContractBlueprint` in the future as well?

Leechael commented:
> * Do you plan to retire `create()` API in the future, given we already have `PinkContractPromise`?
Yup.

> * Will you add `PinkContractBlueprint` in the future as well?
Yup.

Leechael commented:
Added two built-in contract classes: `SystemContractPromise` and `PinkLoggerContractPromise`.

Both of them extend exists `PinkContractPromise` with:
- Hiding unnecessary initial parameters to end-users;
- Typed API (IDE friendly & auto-completion friendly).

A quick examples creating these two contracts:
```typescript
  const api = await ApiPromise.create({
    provider: new WsProvider("wss://poc5.phala.network/ws"),
    types: {
      ...PhalaTypes,
      ...typeDefinitions,
    },
    noInitWarn: true,
  })

  const phatRegistry = await OnChainRegistry.create(api)

  const system = await Phala.SystemContractPromise.create(api, phatRegistry)
  const loggerContractId = await system.getDriver('PinkLogger')
  console.log('Registered logger contract id:', loggerContractId)

  // `system` is optional and it will auto create one if you don't provide one
  const logger = await Phala.PinkLoggerContractPromise.create(api, phatRegistry, system)
  const logs = await logger.getLog(contractId)
  console.log(logs)
```

Leechael commented:
I think it's enough for one PR (CodePromise & BlueprintPromise will be another PR). We need to review the whole design (is it easy to use & easy to understand?), added some tests, and apply the change to e2e tests.

h4x3rotab commented on frontend/packages/sdk/src/contracts/SystemContract.ts:
```diff
@@ -0,0 +1,131 @@
+  }
+
+  // @fixme Not yet test.
+  async stopSidevmAt(contractId: AccountId | string) {
```
I believe those are all not query, but tx.

Leechael replied:
Now remove the `SystemContract` temporary, cause it's not yet ready.

h4x3rotab commented on frontend/packages/sdk/src/contracts/SystemContract.ts:
Do we have any way to reduce the boilerplate code in SystemContract and PinkLogger?

Leechael replied:
I suggest creating a CLI tool similar to phjs that can generate code from ABI files, similar to how phjs generates code from protobuf files.

h4x3rotab replied:
Cool. Let's consider it later.

Leechael replied:
After revisiting for several days, I have come to the conclusion that the generate tools are important. The SystemContract and LoggerContract classes were created as a means of improving DX, but their implementation is still buggy. While we may not highlight these in the first version, they serve as a concept preview.

The most important changes in this version of the SDK should be PinkContractPromise and OnChainRegistry.

h4x3rotab commented on frontend/packages/sdk/src/OnChainRegistry.ts:
```diff
@@ -0,0 +1,110 @@
+      result.forEach(([storageKey, value]) => {
+        const clusterId = storageKey.toHuman()
+      // @ts-ignore
+        const clusterInfo = value.unwrap().toJSON()
```
Why `toJSON()`? If not, you can get the `gasPrice` BN object directly, instead of converting back and forth.

Leechael commented:
Some quick updates:
1. Added `getClusters` and `getWorkers` to `OnChainRegistry` as quick access methods to retrieve related information.
2. Added optional options for `OnChainRegistry.create` that:
  - specified `clusterId` and ensure it's valid.
  - specified Pruntime by `workerId` (the worker pubkey) or `pruntimeURL`, will check it exists or not in connected endpoint.
  - Turn off `autoConnect` (for testnet not yet initialized or some specified scenarios).
3. Explicit `connect` method for `OnChainRegistry` instance.

Quick memo code example grab from the response to the closed-beta feedback:

```
const phatRegistry = await OnChainRegistry.create(api, { pruntimeURL: 'https://phat-fr.01.do/node-1/' })
```

Leechael commented:
Drafting for `PinkCodePromise` and `PinkBlueprintPromise`:

```javascript
// PinkCodePromise. Checking balance in the cluster is optional based on your scenario.
const balance = await phatRegistry.getClusterBalance(user, user.address)
if ((balance.free.toPrimitive() / 1e12) < 1) {
  await signAndSend(phatRegistry.transferToCluster(user.address, 1e12), user)
}
const codePromise = new Phala.PinkCodePromise(api, phatRegistry, contractFile, contractFile.source.wasm)
const { result } = await signAndSend(codePromise.tx.new({}), user)
await result.waitReady(user)

// PinkBlueprintPromise, the eliminating is still in progress.
const blueprintPromise = new Phala.PinkBlueprintPromise(api, phatRegistry, contractFile, contractFile.source.hash)
instantiateResult = await signAndSend(
  blueprintPromise.tx.new({ gasLimit: 1e12, storageDepositLimit: null }),
  user
)
```

Leechael commented:
@kvinwang Please review these two commits, as they may change the behavior: 
- https://github.com/Phala-Network/phala-blockchain/pull/1159/commits/396dc1fd745fbaafbacfbb167fbeb8e19ecd3d8c
- https://github.com/Phala-Network/phala-blockchain/pull/1159/commits/beeba2945601483e515c02c8df5fd8f80d97df7e

h4x3rotab commented on frontend/packages/sdk/src/contracts/PinkBlueprint.ts:
```diff
@@ -0,0 +1,309 @@
+    const { encodedEncryptedData } = res;
+    const { data: encryptedData, iv } = api
+      .createType("EncryptedData", encodedEncryptedData)
+      .toJSON() as {
```
Prefer using the polkadot object directly (`toHex()` or `u8aToHex()` when necessary) than `toJSON()` and `toHuman()`. The latter methods are not lossless.

h4x3rotab commented on e2e/src/setup-drivers.js:
```diff
@@ -44,19 +44,19 @@ async function deployDriverContract(api, txqueue, system, pair, cert, contract,
-    const queryResult = queryResponse.result.toHuman()
-    console.log("InkMessageReturn", queryResult.Ok.InkMessageReturn);
-    const instantiateResult = api.createType('ContractInstantiateResult', queryResult.Ok.InkMessageReturn);
+    // console.log("instantiate result:", instantiateReturn);
```
What's the comment about?

Leechael replied:
Dead codes, relative to this commit: https://github.com/Phala-Network/phala-blockchain/commit/ac1b0ade95a8bb9b11f474184569189e0917eb63

Which move the logic into SDK.

Leechael replied:
I'm not sure this script still in use or not, it's broken as I checked.

h4x3rotab commented on frontend/packages/sdk/src/contracts/PinkContract.ts:
```diff
@@ -0,0 +1,315 @@
+
+
+function createQuery(meta: AbiMessage, fn: (origin: string | AccountId | Uint8Array, options: CertificateData, params: unknown[]) => ContractCallResult<'promise', ContractCallOutcome>): ContractInkQuery<'promise'> {
+  return withMeta(meta, (origin: string | AccountId | Uint8Array, options: CertificateData, ...params: unknown[]): ContractCallResult<'promise', ContractCallOutcome> =>
```
Here, `(origin, option, ...)` becomes `(origin, cert, ...)`. I'm wondering why.

In the previous version, `origin` means the sender. Vanilla ink doesn't have privacy. So everyone can pass the strong of the account id as the origin. In Phat Contract there's privacy and permission. So we pass a signed certificate instead.

In this version, do we need to pass both the origin and the certificate? Where's the options (to specify gas fee, etc)?

Leechael replied:
The original Polkadot contract query API includes options such as `gasLimit` and `storageDepositLimit`, which are irrelevant for off-chain queries. Instead, users must provide a certificate. To reflect this change, we propose replacing the `options` parameter with `cert`. Additionally, we suggest making the handling of certificate caching user-controlled rather than internal. Initially, we attempted to handle certificate signing automatically but found that it caused pop-ups in the browser asking users to sign.

```
contractPromise.query.foo(address, options, ....args)
// Become
contractPromise.query.foo(address, cert, ....args)
```

h4x3rotab replied:
In the context of Phat Contract, the original `address` is irrelevant. The cert represents the identity. So I think it makes more sense to replace address by the cert. On the other hand, the options is not used. So probably we can just leave it as an empty struct. Sample would look like:

```typescript
interface PhatContractOption {}
contractPromise.query.foo(cert, {}, ....args)
```

Leechael replied:
> In the context of Phat Contract, the original `address` is irrelevant. The cert represents the identity. So I think it makes more sense to replace address by the cert. On the other hand, the options is not used. So probably we can just leave it as an empty struct. Sample would look like:
> 
> ```ts
> interface PhatContractOption {}
> contractPromise.query.foo(cert, {}, ....args)
> ```

This is what current API looks like:

```ts
const aliceCert = await signCertificate({ pair: alice, api })
const { gasRequired, storageDeposit } = await contractPromise.query.newBadge(alice, aliceCert, 'Badge#1')
const options = {
   gasLimit: gasRequired.refTime,
   storageDepositLimit: storageDeposit.isCharge ? storageDeposit.asCharge : null,
}
await contractPromise.tx.newBadge(options, 'Badge#1').signAndSend(alice)
```

If we skip the address:

```ts
const aliceCert = await signCertificate({ pair: alice, api })
const { gasRequired, storageDeposit } = await contractPromise.query.newBadge(aliceCert, {}, 'Badge#1')
const options = {
    gasLimit: gasRequired.refTime,
    storageDepositLimit: storageDeposit.isCharge ? storageDeposit.asCharge : null,
}
await contractPromise.tx.newBadge(options, 'Badge#1').signAndSend(alice)
```

The original API:

```ts
const { gasRequired, storageDeposit, result } = await contract.query.newBadge(alice, { storageDepositLimit: null, gasLimit: -1 }, 'Badge#1');
const options = {
    gasLimit: gasRequired.refTime,
    storageDepositLimit: storageDeposit.isCharge ? storageDeposit.asCharge : null,
}
await contract.tx.newBadge(options, 'Badge#1').signAndSend(alice)
```

Or we can re declare the options as:
```ts
interface PhatContractQueryOption {
  cert: any
}

const queryOptions = await Phala.getQueryOptions(alice)
const { gasRequired, storageDeposit } = await contractPromise.query.newBadge(alice, options, 'Badge#1')
const txOptions = {
    gasLimit: gasRequired.refTime,
    storageDepositLimit: storageDeposit.isCharge ? storageDeposit.asCharge : null,
}
await contractPromise.tx.newBadge(txOptions, 'Badge#1').signAndSend(alice)
```

Then we can get more similar API.

Leechael replied:
Let me give another example show the benefit of keeping similar API:

```ts
const api = ApiPromise.create({ ... })
const DIProvider = new PolkadotCompatibleContractServiceProvider(api)
//
// We don't care about it's original Polkadot JS ContractPromise instance or PinkContractPromise instance
//
const contractPromise = await DIProvider.createContractPromise(contractId)
const queryOptions = awiat DIProvider.createContractQueryOptions(alice)
await contractPromise.query.newBadge(alice, queryOptions, 'Badge #1')
```
